x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/editor_frame.tsx

```diff
@@ -35,28 +36,78 @@ export function EditorFrame(props: EditorFrameProps) {
   const [state, dispatch] = useReducer(reducer, props, getInitialState);
   const { onError } = props;
 
-  // create public datasource api for current state
-  // as soon as datasource is available and memoize it
-  const datasourcePublicAPI = useMemo(
-    () =>
-      state.datasource.activeId && !state.datasource.isLoading
-        ? props.datasourceMap[state.datasource.activeId].getPublicAPI(
-            state.datasource.state,
-            (newState: unknown) => {
+  const allLoaded = Object.values(state.datasourceStates).every(
+    ({ isLoading }) => typeof isLoading === 'boolean' && !isLoading
+  );
+
+  // Initialize current datasource and all active datasources
+  useEffect(() => {
+    if (!allLoaded) {
+      Object.entries(props.datasourceMap).forEach(([datasourceId, datasource]) => {
+        if (
+          state.datasourceStates[datasourceId] &&
+          state.datasourceStates[datasourceId].isLoading
+        ) {
+          datasource
+            .initialize(state.datasourceStates[datasourceId].state || undefined)
+            .then(datasourceState => {
               dispatch({
                 type: 'UPDATE_DATASOURCE_STATE',
-                newState,
+                newState: datasourceState,
+                datasourceId,
               });
-            }
-          )
-        : undefined,
-    [
-      props.datasourceMap,
-      state.datasource.isLoading,
-      state.datasource.activeId,
-      state.datasource.state,
-    ]
-  );
+            })
+            .catch(onError);
+        }
+      });
+    }
+  }, [allLoaded]);
+
+  const datasourceLayers: Record<string, DatasourcePublicAPI> = {};
+  Object.keys(props.datasourceMap).forEach(id => {
+    const stateWrapper = state.datasourceStates[id];
+    if (!stateWrapper || stateWrapper.isLoading) {
+      return;
+    }
+    const dsState = stateWrapper.state;
+    const layers = props.datasourceMap[id].getLayers(dsState);
+
+    layers.forEach(layer => {
+      const publicAPI = props.datasourceMap[id].getPublicAPI(
+        dsState,
+        (newState: unknown) => {
+          dispatch({
+            type: 'UPDATE_DATASOURCE_STATE',
+            datasourceId: id,
+            newState,
+          });
+        },
+        layer
+      );
+
+      datasourceLayers[layer] = publicAPI;
+    });
+  });
+
+  const framePublicAPI: FramePublicAPI = {
+    datasourceLayers,
+    addNewLayer: () => {
+      const newLayerId = generateId();
+
+      const newState = props.datasourceMap[state.activeDatasourceId!].insertLayer(
```

Comment:
I don't love that this is effectful, but I understand why it is. Right now, I'm calling addNewLayer in visualization initialization, as we need a layer in order to have a valid state. It feels messy that that causes an effect like this. Also, it seems we could conceivably have a scenario in which a visualization's suggestions might want to create more than one layer, though not sure about that.

Probably not a blocker for merging, but it's worth noting.

If we could make this work similarly to how adding a new accessor works, that would be nice.

Reply:
Good point, this is a general problem with our state handling - multiple (state changing) calls to the frame/datasourceapi in the same synchronous stack would overwrite each other. This is what I meant in the chat yesterday.